### PR TITLE
RESTWS-562 Improve Resource Definition Documentation

### PR DESCRIPTION
**Pull request by gayanW**
- adds swagger-core library
  * uses its Java API and model classes to construct and generate the swagger json.
- introudce new methods which model objects used in constructing the definitions section:
  * getGETModel(Representation) :  returns a object represting GET representation schema of resource
  * getCREATEModel(Representation) : returns a object representing CREATE representation shema of a resource
  * getUPDATEModel(Representation) : returns a object representing POST Update representation schema of a resource

ex: PersonGet, PersonCreate, PersonUpdate
- adds new test cases to: SwaggerSpecificationCreatorTest.java

- adds support for multiple representation types
GET schemas have support for representations: DEFAULT, REF, and FULL
CREATE schemas have support for representations: DEFAULT, and FULL
POST_UPDATE schemas have support for representations: DEFAULT

  * uses default schema for the shema of responses and parameters

Swagger before: http://www.jsoneditoronline.org/?id=6ac2d0c797f5e8e4060d09b62b300b04
Swagger after: http://jsoneditoronline.org/?id=d692754b9b4543f7388572317dc10400

### CODE REVIEW

**Comment by dkayiwa** [comment]
@gayanW did you see the travis failure?

**Comment by gayanW** [comment]
@dkayiwa minor mistake : ) please check now

**Comment by psbrandt** [comment]
@bholagabbar, are you reviewing this PR?

For this to work, do developers have to implement the `getGETModel`, `getCREATEModel` and `getUPDATEModel` methods in their resources? I haven't taken a look at the code, but is this enforced somewhere? What happens if these methods aren't implemented?

**Comment by gayanW** [comment]
@psbrandt 
> For this to work, do developers have to implement the getGETModel, getCREATEModel and getUPDATEModel methods in their resources? I haven't taken a look at the code, but is this enforced somewhere? What happens if these methods aren't implemented?

Unless introducing a new resource, it is not mandatory to implement those methods. However if not, their definitions won't appear in the swagger doc. But it will not affect the program in any other way.

I've updated almost all the resources within the rest module. Methods  such as getRepresentationDescription(..) is no longer being used to generate the documentation. However I did not remove any of these existing code in case they are being accessed by other modules.

**Comment by bholagabbar** [comment]
@psbrandt @dkayiwa Sorry, I've been away on the Google trip. I am trying to make an attempt to yes, but reviewing 117 files changes !? There probably was a better way to break it into subtasks

File: omod-common/src/main/java/org/openmrs/module/webservices/docs/swagger/SwaggerSpecificationCreator.java [hunk 1]
```diff
@@ -9,20 +9,15 @@
...
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Map.Entry;
-import java.util.Set;
-
+import io.swagger.models.*;
```
Comment (bholagabbar) [inline omod-common/src/main/java/org/openmrs/module/webservices/docs/swagger/SwaggerSpecificationCreator.java]: @gayanW No wildcard imports. Period. Remove all of them and replace them with individual classes

File: omod-common/pom.xml [hunk 2]
```diff
@@ -29,6 +29,22 @@
...
 		</dependency>
+
+		<dependency>
+			<groupId>io.swagger</groupId>
+			<artifactId>swagger-core</artifactId>
+			<version>1.5.13</version>
+            <exclusions>
+                <exclusion>
+					<groupId>org.slf4j</groupId>
+					<artifactId>slf4j-api</artifactId>
```
Comment (bholagabbar) [inline omod-common/pom.xml]: Error in XML formatting. Indentation is off

File: omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java [hunk 3]
```diff
@@ -93,6 +97,36 @@ public DelegatingResourceDescription getUpdatableProperties() throws ResourceDoe
...
+		ModelImpl model = (ModelImpl) super.getGETModel(rep);
+		if (rep instanceof DefaultRepresentation || rep instanceof FullRepresentation) {
+			model
+			        .property("uuid", new StringProperty())
+			        .property("display", new StringProperty())
+			        .property("startDate", new DateProperty())
+			        .property("endDate", new DateProperty())
+			        .property("patientUuid", new StringProperty());
+		}
+		//FIXME missing props
```
Comment (bholagabbar) [inline omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java]: ?

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/ProviderAttributeTypeResource1_9.java [hunk 4]
```diff
@@ -9,12 +9,14 @@
  */
 package org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_9;
 
+import io.swagger.models.Model;
```
Comment (bholagabbar) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/ProviderAttributeTypeResource1_9.java]: Is this import being used here? Get rid of all unused imports in all the files if they exist

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java [hunk 5]
```diff
@@ -157,10 +192,10 @@ public void checkPagingParamsExist() throws Exception {
 		}
 	}
 	
-	private boolean operationHasPagingParams(Operation o) {
+	private boolean operationHasPagingParams(io.swagger.models.Operation o) {
 		boolean limit = false, startIndex = false;
 		
-		for (Parameter p : o.getParameters()) {
+		for (io.swagger.models.parameters.Parameter p : o.getParameters()) {
```
Comment (gayanW) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: I'll replace all possible fully qualified class names.

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/AnimalClassResource_1_9.java [hunk 6]
```diff
@@ -9,6 +9,7 @@
  */
```
Comment (gayanW) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/AnimalClassResource_1_9.java]: They have always been there. I only implemented the missing methods. It seems that they are only used by `RestServiceImplTest`

Comment (bholagabbar) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/AnimalClassResource_1_9.java]: @dkayiwa What do you think? Should we have these tests? On the surface, `AnimalClassResource` doesn't seem right inside an OpenMRS webservices module. Ideally we should be replacing these tests?

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 5 above)
Comment (gayanW) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: Done

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/ProviderAttributeTypeResource1_9.java (hunk 4 above)
Comment (gayanW) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/ProviderAttributeTypeResource1_9.java]: Done

**Comment by gayanW** [comment]
FYI: IMO reviewing all resource classes is not going to be practical. It takes a lot of effort to find right attributes, and their respective types of a resource. I had to go through ones' parent classes, supportedClass, respective tests and so on. So I think It will suffice therefore to review  SwaggerSpecificationCreator and the generated json spec. And may be just go through the resource classes.

**Comment by bholagabbar** [comment]
@dkayiwa @psbrandt I've asked gayan to push commits as he fixes issues so it makes it easier for us to review PRs. Once we review everything and give him the go ahead, he can squash and then merge. Reviewing 280 files changes is insanity!

**Comment by dkayiwa** [comment]
@gayanW did you notice the merge conflicts on this?

**Comment by gayanW** [comment]
@dkayiwa The pr was gone stale. I just rebased it.

**Comment by dkayiwa** [comment]
@gayanW your pull request deals with a lot of things, formatting, imports, classname changes, etc.
I would recommend a separate pull request for each of these. That way, we do not have to deal with a huge pull request whose changes are unrelated.

**Comment by gayanW** [comment]
@bholagabbar @dkayiwa 
Thanks for the inputs :)

I squashed, and eliminated those formats/fix done on classes that are not directly relevant to our use case. So now we have only the changes made to the SwaggerSpecificationCreator and the resource classes.

To my understanding, it's inevitable not to update all the resources, because then it would result in a partial swagger spec.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java [hunk 7]
```diff
@@ -102,6 +108,43 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
...
+		if (rep instanceof FullRepresentation) {
+			model
+			        .property("encounter", new RefProperty("#/definitions/EncounterCreate"))
+			        .property("patient", new RefProperty("#/definitions/PatientCreate"))
+			        .property("concept", new RefProperty("#/definitions/ConceptCreate"))
+			        .property("orderer", new RefProperty("#/definitions/UserCreate"))
+			        .property("previousOrder", new RefProperty("#/definitions/OrderCreate"))
+			        .property("orderReason", new RefProperty("#/definitions/ConceptCreate"));
+		}
+		//FIXME missing prop: type
```
Comment (gayanW) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: In 1_8,  `orderType` is marked required. But in 1_10 it seems that this property is no longer required to place a new order. For those that I thought would need extra testing, and research, I put FIXME. May we fix these in a new PR?

Comment (bholagabbar) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Mention this in the next issue you create where you will be fixing the changes you removed formatting, etc not directly relevant to this issue

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java [hunk 8]
```diff
@@ -102,6 +108,43 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
 		}
 	}
 	
+	@Override
+	public Model getCREATEModel(Representation rep) {
```
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Did you intentionally name this as getCREATEModel instead of getCreateModel?

File: OpenMRSFormatter.xml [hunk 9]
```diff
@@ -263,5 +263,6 @@
 <setting id="org.eclipse.jdt.core.formatter.insert_space_before_opening_angle_bracket_in_type_arguments" value="do not insert"/>
 <setting id="org.eclipse.jdt.core.formatter.never_indent_block_comments_on_first_column" value="false"/>
 <setting id="org.eclipse.jdt.core.formatter.keep_then_statement_on_same_line" value="false"/>
+<setting id="org.eclipse.jdt.core.formatter.join_wrapped_lines" value="false"/>
```
Comment (dkayiwa) [inline OpenMRSFormatter.xml]: The above is an example of a change that should have been in a separate pull request.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/FieldTypeResource1_8.java [hunk 10]
```diff
@@ -29,6 +32,25 @@
...
+		if (rep instanceof DefaultRepresentation || rep instanceof FullRepresentation) {
+			modelImpl
+			        .property("isSet", new BooleanProperty()._default(false));
+		}
+		return modelImpl;
+	}
+	
+	@Override
+	public Model getCREATEModel(Representation rep) {
+		return super.getCREATEModel(rep); //FIXME missing props
```
Comment (dkayiwa) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/FieldTypeResource1_8.java]: Are the FIXME supposed to be addressed in another pull request?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java [hunk 11]
```diff
@@ -74,7 +79,7 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
 			description.addProperty("validator");
 			description.addProperty("locationBehavior");
 			description.addProperty("uniquenessBehavior");
-			description.addProperty("validator");
+			description.addProperty("validator"); //FIXME duplicate
```
Comment (dkayiwa) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java]: When you do plan to fix the above?

File: omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java (hunk 3 above)
Comment (dkayiwa) [inline omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java]: @gayanW did you see the above question mark?

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 8 above)
Comment (gayanW) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: May be I thought that it was more readable. Anyway I could change that.

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java [hunk 12]
```diff
@@ -9,31 +9,73 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static junit.framework.TestCase.assertNotNull;
+
 public class SwaggerSpecificationCreatorTest extends BaseModuleWebContextSensitiveTest {
 	
+	@Test
+	public void mainTest() {
+		String baseUrl = "localhost:8080/openmrs/ws/rest";
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: Does this work even when the 8080 port is used by some other process?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/FieldTypeResource1_8.java (hunk 10 above)
Comment (gayanW) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/FieldTypeResource1_8.java]: Yes. As discussed with Shreyans they will be fixed in a separate issue. These will require minor to no change at all. But may need testing and discussion.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 8 above)
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Do your mentors prefer getCREATEModel to getCreateModel?

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 12 above)
Comment (gayanW) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: I just tested this with different listening and open ports. And it has no effect on where the API is exposed. It changes the baseUrl of the swagger json, and as this is a test it has no much significance.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 8 above)
Comment (gayanW) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: I don't think they commented on this. If you prefer `getCreateModel` over `getCREATEModel`, I can refactor.

File: omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java (hunk 3 above)
Comment (gayanW) [inline omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java]: @dkayiwa IMO. lines marked FIXME that'll update the swagger properties will be better fix in a new separate issue.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 8 above)
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Can you ask them and tell me what they prefer?

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 12 above)
Comment (bholagabbar) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: In either case, hard coding a port number is not good practice. Make BaseUrl generic if it doesn't make a difference

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java (hunk 11 above)
Comment (bholagabbar) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java]: @gayanw create an issue on JIRA and paste a link here for reference

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java (hunk 8 above)
Comment (bholagabbar) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Since GET, CREATE and UPDATE are actions that  I are being documented, captial cases indicating  the action makes sense. I understand the deviation from camelCase but this makes the code more readable. I am fine with this. @dkayiwa I leave the final decision upto you

Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: I just wanted to make sure that you are aware of this. 😊

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 12 above)
Comment (gayanW) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: The port or the baseUrl being passed in here have no actual significance. As you could see with old tests: https://github.com/gayanW/openmrs-module-webservices.rest/blob/281145120aee4664d048b0b1cb2f6264828d2035/omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java#L142

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java (hunk 11 above)
Comment (gayanW) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java]: @bholagabbar should I create, before or after this is being merged?

Comment (bholagabbar) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java]: create it now and paste the link, is there an issue with this?

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 12 above)
Comment (bholagabbar) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: Hmm alright. I'm convinced when you tell me that it has no significance/doesn't affect the code but having something like 'http:8080/localhost' is never good practice even if it's in tests. Just assign it as 'base_url/ws/rest` or something if it doesn't have any significance.

Comment (gayanW) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: Changed to "host/openmrs/ws/rest"
FYI: https://swagger.io/docs/specification/2-0/api-host-and-base-path/

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java (hunk 11 above)
Comment (gayanW) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierTypeResource1_8.java]: https://issues.openmrs.org/browse/RESTWS-676

File: omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java (hunk 3 above)
Comment (gayanW) [inline omod-2.1/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_1/CohortMembershipResource2_1.java]: Will be fixed in: https://issues.openmrs.org/browse/RESTWS-676

File: omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java (hunk 12 above)
Comment (bholagabbar) [inline omod/src/test/java/org/openmrs/module/webservices/rest/doc/SwaggerSpecificationCreatorTest.java]: Ok this looks good

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java [hunk 13]
```diff
@@ -60,6 +61,21 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
 		return null;
 	}
 	
+	@Override
+	public Model getGETModel(Representation representation) {
+		return null;
+	}
+
```
Comment (dkayiwa) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java]: I see a number of these methods simply return null. Shouldn't it be enough not to have them and assume that whichever class does not have them, null it is? Am looking at reducing the amount of code that we have to maintain.

**Comment by dkayiwa** [comment]
The ticket status also says that more work is still going on.

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java (hunk 13 above)
Comment (gayanW) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java]: @dkayiwa 
Yeah I see. But isn't that good that we enforce them? So when creating a resource class one knows that he should document them? BTW I'm following a similar approach as it's done in old getRepresentationDescription

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/DuplicateNameAnimalResource_1_9.java [hunk 14]
```diff
@@ -60,6 +61,21 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
 		return null;
 	}
 	
+	@Override
+	public Model getGETModel(Representation representation) {
+		return null;
+	}
+
```
Comment (dkayiwa) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/DuplicateNameAnimalResource_1_9.java]: Are these new methods also fully documented anywhere in the code? Am looking at it from the point of the maintainers who will look at these new methods and try to figure out why they were added.

Comment (gayanW) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/DuplicateNameAnimalResource_1_9.java]: Yeah it's necessary that I write a documentation page, explaining how to document when introducing a new resource class. I will do that.

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java (hunk 13 above)
Comment (gayanW) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java]: @dkayiwa It's your call :) ideally only a test resource should return null values. Other resource will either implement or could just inherit from its parent

Comment (dkayiwa) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java]: In which class are these methods declared?

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/DuplicateNameAnimalResource_1_9.java (hunk 14 above)
Comment (dkayiwa) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/DuplicateNameAnimalResource_1_9.java]: I mean documentation that is not outside the source 😊

File: omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java (hunk 13 above)
Comment (gayanW) [inline omod-common/src/test/java/org/mockingbird/test/rest/resource/BirdResource_1_9.java]: Interface: https://github.com/gayanW/openmrs-module-webservices.rest/blob/270edab5222fdfd7b8a77adf209e6d280c1688af/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/DelegatingResourceHandler.java#L93-L97

Almost all resources inherit: https://github.com/gayanW/openmrs-module-webservices.rest/blob/270edab5222fdfd7b8a77adf209e6d280c1688af/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/BaseDelegatingResource.java#L76-L101

Yeah I just saw a minor improvement that could be done. I'll push it with the added doc.

**Comment by dkayiwa** [comment]
@gayanW can you also edit the ticket description with a detailed description of the approach you decided to implement in this pull request as the way to improve the resource definition documentation?

**Comment by gayanW** [comment]
I documented the methods. see: https://github.com/gayanW/openmrs-module-webservices.rest/blob/b9a970ed4ec1c1696410e24489c3445181a7aca1/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/DelegatingResourceHandler.java#L93-L126

And updated the issue description as well.
@dkayiwa @bholagabbar

**Comment by dkayiwa** [comment]
@gayanW i have tried running these changes locally and saw a few more resources show up like "address", "allergy", "answer". Was this an intentional side effect of this pull request? Or is it something else?

**Comment by dkayiwa** [comment]
@gayanW can you engage the community to see what others say?

**Comment by gayanW** [comment]
I changed it to the old way. (FYI: only these 2 lines were changed - [SwaggerSpecificationCreator.java#L602](https://github.com/gayanW/openmrs-module-webservices.rest/blob/07897dd5238fd103e75ab0747acafed1f0e51c46/omod-common/src/main/java/org/openmrs/module/webservices/docs/swagger/SwaggerSpecificationCreator.java#L602) [SwaggerSpecificationCreator.java#L1001](https://github.com/gayanW/openmrs-module-webservices.rest/blob/07897dd5238fd103e75ab0747acafed1f0e51c46/omod-common/src/main/java/org/openmrs/module/webservices/docs/swagger/SwaggerSpecificationCreator.java#L1001))

**Comment by dkayiwa** [comment]
Ok let me compile and test it out again.

**Comment by gayanW** [comment]
@dkayiwa 
You may also have a look at the final json: https://<>/openmrs/module/webservices/rest/swagger.json

The changes will be visible once we upgrade the ui: [RESTWS-560](https://issues.openmrs.org/browse/RESTWS-560)

**Comment by dkayiwa** [comment]
So how will i see the changes before RESTWS-560 is done?

**Comment by gayanW** [comment]
@dkayiwa Some changes are still visible with the old ui. But if you want to see all the changes, you may try this out: http://editor.swagger.io/

And paste the content of the swagger.json there.

**Comment by dkayiwa** [comment]
How long would it take to finish the other ticket?

**Comment by gayanW** [comment]
@dkayiwa It is better to work on it once this is finished.

**Comment by dkayiwa** [comment]
@gayanW i have just tested again and noticed that the q parameter is not highlighted as required, with your changes. Was that intentional? https://qa-refapp.openmrs.org/openmrs/module/webservices/rest/apiDocs.htm#!/order/getAllOrders

**Comment by gayanW** [comment]
@dkayiwa 
Are you sure it is not marked required for `GET /order`
http://localhost:8080/openmrs/module/webservices/rest/apiDocs.htm#!/order/getAllOrders
I ran locally and it is marked required.

q param is not marked required for resources that supports fetchAll.
`/concept    Fetch all non-retired concept resources or perform search`
http://localhost:8080/openmrs/module/webservices/rest/apiDocs.htm#!/concept/getAllConcepts

If there's more than one query param, then it is not marked required either.
http://localhost:8080/openmrs/module/webservices/rest/apiDocs.htm#!/encounter/getAllEncounters

**Comment by dkayiwa** [comment]
@bholagabbar what behavior do you get in regards to the above?

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java [hunk 15]
```diff
@@ -102,6 +108,43 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
...
+		ModelImpl model = new ModelImpl()
+		        .property("encounter", new StringProperty().example("uuid"))
+		        .property("action", new StringProperty()
+		                ._enum(SwaggerSpecificationCreator.getEnumsAsList(Order.Action.class)))
+		        .property("accessionNumber", new StringProperty())
+		        .property("dateActivated", new DateProperty())
+		        .property("scheduledDate", new DateProperty())
+		        .property("patient", new StringProperty().example("uuid"))
```
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: Where are these examples displayed?

Comment (gayanW) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/OrderResource1_10.java]: http://localhost:8080/openmrs/module/webservices/rest/apiDocs.htm#!/order/createOrder
See in the 'Example Value' section.

{
  "encounter": "uuid",
  "action": "NEW",
  "accessionNumber": "string",
  "dateActivated": "2017-08-24",
  "patient": "**uuid**",
  "concept": "uuid",
...
}

File: omod-1.11/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_11/DrugIngredientResource1_11.java [hunk 16]
```diff
@@ -80,6 +81,21 @@ public DelegatingResourceDescription getUpdatableProperties() {
 		return getCreatableProperties();
 	}
 	
+	@Override
+	public Model getGETModel(Representation representation) {
+		return null;
+	}
+
```
Comment (dkayiwa) [inline omod-1.11/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_11/DrugIngredientResource1_11.java]: What is the difference between a resource like this which returns all nulls and those that do not? Or was it simply a lack of time to implement?

Comment (gayanW) [inline omod-1.11/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_11/DrugIngredientResource1_11.java]: No it's not due to lack of time. I had enough time :) I've been tracking undocumented resources by validating the swagger. Some how they were missed. I'll check and fix those.

Comment (dkayiwa) [inline omod-1.11/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_11/DrugIngredientResource1_11.java]: And i found a good number of them.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java [hunk 17]
```diff
@@ -123,6 +124,21 @@ public DelegatingResourceDescription getUpdatableProperties() throws ResourceDoe
 		return orderResource.getUpdatableProperties();
 	}
 	
+	@Override
+	public Model getGETModel(Representation representation) {
+		return null;
+	}
+	
+	@Override
```
Comment (dkayiwa) [inline omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java]: Were these intentionally left null?

File: omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/OrderSetMemberResource1_12.java [hunk 18]
```diff
@@ -99,6 +100,16 @@ public DelegatingResourceDescription getUpdatableProperties() {
 	}
 	
 	@Override
+	public Model getGETModel(Representation representation) {
+		return null;
+	}
```
Comment (dkayiwa) [inline omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/OrderSetMemberResource1_12.java]: Were these intentionally left null? And many more others.

File: omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java (hunk 17 above)
Comment (gayanW) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java]: Do we need to document handlers? I don't think they appear either as a resource or a property. If needed let me know.

Comment (dkayiwa) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java]: Then in that case we should not have these methods in them.

Comment (gayanW) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java]: However they implements getRepresentationDescription, getCreatableProperties, getCreatableProperties. Is handler considered a resource or a property of some other resource?

Comment (dkayiwa) [reply omod-1.10/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_10/TestOrderSubclassHandler1_10.java]: I would not consider it as a resource. But putting this aside, can you ensure that all other non search handlers that should not just return null, are doing what they should?

**Comment by gayanW** [comment]
@dkayiwa 
ModuleResource1_8, OrderableResource1_10, CustomDatatypeResource1_9, ConceptSearchResource1_9, and CustomDatatypeHandlerResource1_9  returns null for getCREATEModel(rep) because they are documented as not creatable. If you check these classes, you'll see that none of them implement getCreatableProperties() and throws ResourceDoesNotSupportOperationException when invoked.

Test resource classes also returns null, as they are not needed.

Except these, all other classes are documented.

**Comment by dkayiwa** [comment]
Ok cool. 
Am merging this earlier than later such that you get un blocked.
After merging, i will take a second look and if any more work is needed, you will do that in a new pull request. Otherwise thanks for the work so far done. 😊
